pulsar-broker/src/main/java/org/apache/pulsar/compaction/Compactor.java

```diff
@@ -0,0 +1,205 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.compaction;
+
+import com.google.common.collect.ImmutableMap;
+import io.netty.buffer.ByteBuf;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.bookkeeper.client.BKException;
+import org.apache.bookkeeper.client.BookKeeper;
+import org.apache.bookkeeper.client.LedgerHandle;
+
+import org.apache.pulsar.common.api.Commands;
+import org.apache.pulsar.common.api.proto.PulsarApi.MessageMetadata;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.RawReader;
+import org.apache.pulsar.client.api.RawMessage;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Compactor for Pulsar topics
+ *
+ * Compaction will go through the topic in two passes. The first pass
+ * selects latest offset for each key in the topic. Then the second pass
+ * writes these values to a ledger.
+ *
+ * <p>The two passes are required to avoid holding the payloads of each of
+ * the latest values in memory, as the payload can be many orders of
+ * magnitude larger than a message id.
+*/
+public class Compactor {
+    private static final Logger log = LoggerFactory.getLogger(Compactor.class);
+    private static final String COMPACTION_SUBSCRIPTION = "__compaction";
+    private static final int MAX_OUTSTANDING_READS = 500;
+    private static final int COMPACTED_TOPIC_ENSEMBLE_SIZE = 3;
+    private static final int COMPACTED_TOPIC_ACK_QUORUM_SIZE = 3;
+    private static final String COMPACTED_TOPIC_LEDGER_PROPERTY = "CompactedTopicLedger";
+    static byte[] COMPACTED_TOPIC_LEDGER_PASSWORD = COMPACTION_SUBSCRIPTION.getBytes();
+
+    private final ExecutorService executor;
+    private final PulsarClient pulsar;
+    private final BookKeeper bk;
+
+    public Compactor(PulsarClient pulsar,
+                     BookKeeper bk,
+                     ExecutorService executor) {
+        this.executor = executor;
+        this.pulsar = pulsar;
+        this.bk = bk;
+    }
+
+    public CompletableFuture<Long> compact(String topic) {
+        final CompletableFuture<Long> future
+            = new CompletableFuture<>();
+        executor.submit(new CompactionTask(topic, future));
+        return future;
+    }
+
+    private class CompactionTask implements Runnable {
```

Comment:
nit: it might be better to consider writing this in asynchronous, it should be not difficult using complete future chains. this can be a future improvement.

Reply:
Are you suggesting something like:
```
while (/* something */) {
    reader.readNext().theCompose((m) -> {
         CompletableFuture<?> future = new CompletableFuture<>();
         lh.asyncAddEntry(process(m),
                   (rc, ledger, eid, ctx) -> { /* blah blah */));
         return future;
     }).
```

How do you limit the number of outstanding reads? How would you break out of the loop?

Reply:
Actually, I guess you mean something like:
```
CompletableFuture<?> readOne() {
    return reader.readNext().theCompose((m) -> {
         CompletableFuture<?> future = new CompletableFuture<>();
         lh.asyncAddEntry(process(m),
                   (rc, ledger, eid, ctx) -> { /* blah blah */));
         return readOne();
     });
}
```

I'll play around a bit, see if I can make it work.

Reply:
Actually, this would result in a load of futures in memory (as many as there are messages in the topic), which wouldn't be good.

Reply:
Without a synchronous .get(timeout, unit), how would we detect we had hit the end of the ledger?

Reply:
The more I think about it, I don't think a async version is possible without a reliable way to know you've reached the end of the topic, which we won't have until #937 complete. I guess we can do it once #1066 is in.

Reply:
@ivankelly 

1) this is not a strong opinion. it is okay to start with a sync version

2) in this two pass algorithm, you don't need to do rate limiting on the first pass. you might need to do rate limiting on the second pass, to avoid sending too fast to a ledger. you want can basically need an async semaphore with proper waiting strategy. this [AsyncSemaphore] (https://github.com/apache/distributedlog/blob/master/distributedlog-common/src/main/java/org/apache/distributedlog/common/concurrent/AsyncSemaphore.java) can give some idea. it was a port from twitter scala future.

3) 

> this would result in a load of futures in memory (as many as there are messages in the topic), which wouldn't be good.

using sync version is effectively has same allocation pattern using async version. because all sync version is based on async calls. I don't see how it is going to be better using sync calls.

4) 

> I don't think a async version is possible without a reliable way to know you've reached the end of the topic, which we won't have until #937 complete. 

you don't necessarily need to be blocked on #937. [FutureUtils#within](https://github.com/apache/bookkeeper/blob/master/bookkeeper-common/src/main/java/org/apache/bookkeeper/common/concurrent/FutureUtils.java#L271) is an idea for an equvialent of async version of your current approach. again this util is also a port from twitter scala future.

Reply:
I'll leave moving to async until #937 is in. It looks like that's going to take longer than I expected.

> using sync version is effectively has same allocation pattern using async version. because all sync version is based on async calls. I don't see how it is going to be better using sync calls.

Depends on how you implement it, but in my head it was a bunch of thenCompose calls, which would end up with a bunch of futures holding references to more future and so on. Anyhow, I'll look more into making it async when we have #937.